snapshots/handler/handler.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/identity"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+const (
+	TargetSnapshotLabel = "containerd.io/snapshot.ref"
+	TargetRefLabel      = "containerd.io/snapshot/target.reference"
+	TargetDigestLabel   = "containerd.io/snapshot/target.digest"
+)
+
+// FilterLayerBySnapshotter filters out layers from download candidates if the
+// snapshotter can prepare the layer snapshot without pulling it.
+func FilterLayerBySnapshotter(f images.HandlerFunc, sn snapshots.Snapshotter, store content.Store, fetcher remotes.Fetcher, ref string) images.HandlerFunc {
+	return func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		children, err := f(ctx, desc)
+		if err != nil {
+			return nil, err
+		}
+
+		if desc.MediaType == ocispec.MediaTypeImageManifest ||
+			desc.MediaType == images.MediaTypeDockerSchema2Manifest {
+			p, err := content.ReadBlob(ctx, store, desc)
+			if err != nil {
+				return nil, err
+			}
+			var manifest ocispec.Manifest
+			if err := json.Unmarshal(p, &manifest); err != nil {
+				return nil, err
+			}
+
+			if _, err := remotes.FetchHandler(store, fetcher)(ctx, manifest.Config); err != nil {
+				return nil, err
+			}
+			chain, err := images.RootFS(ctx, store, manifest.Config)
+			if err != nil {
+				return nil, err
+			}
+
+			// Get the index of a layer which is the lowest non-skippable layer. We avoid
+			// downloading skippable layers which are lower than it by filtering out them.
+			i := firstFalseIdx(manifest.Layers, chain, isPullSkippable(ctx, sn, ref))
+			children = exclude(children, manifest.Layers[:i])
+		}
+
+		return children, nil
+	}
+}
+
+func isPullSkippable(ctx context.Context, sn snapshots.Snapshotter, ref string) func(ocispec.Descriptor, []digest.Digest) bool {
+
+	// Check if the given layer is a pull-skippable snapshot. Using chain, this
+	// function generates chainID to prepare/commit the snapshot so that the
+	// snapshot can be Prepare()ed later using the chainID as a normal way.
+	return func(layer ocispec.Descriptor, chain []digest.Digest) bool {
+		var (
+			parent      = identity.ChainID(chain[:len(chain)-1]).String()
+			chainID     = identity.ChainID(chain).String()
+			layerDigest = layer.Digest.String()
+		)
+
+		// If this layer can be provided without pulling the actual contents, the
+		// snapshotter gives us ErrAlreadyExists.
+		for {
+			pKey := getUniqueKey(ctx, sn, chainID)
+			if _, err := sn.Prepare(ctx, pKey, parent, snapshots.WithLabels(map[string]string{
+				TargetSnapshotLabel: chainID,
+				TargetRefLabel:      ref,
+				TargetDigestLabel:   layerDigest,
+			})); !errdefs.IsAlreadyExists(err) {
+				sn.Remove(ctx, pKey)
+				return false
+			}
+			if _, err := sn.Stat(ctx, pKey); err == nil {
```

Comment:
I think we should stat the `chainID` first before the for loop. If the chainID exists, it seems it will be a dead end.